### PR TITLE
Remove cannot mutate statics in initializer of another static error

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk

This is just a refactoring. As the removed code itself said, it only a heuristic catching a few cases early instead of leaving it all to const eval. It's easy to work around the static check and then run into the miri-engine check.

### CODE REVIEW

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -782,19 +782,6 @@ impl<'a, 'tcx> Visitor<'tcx> for Checker<'a, 'tcx> {
 
                 // Only allow statics (not consts) to refer to other statics.
                 if self.mode == Mode::Static || self.mode == Mode::StaticMut {
-                    if self.mode == Mode::Static
```
Comment (oli-obk) [inline src/librustc_mir/transform/qualify_consts.rs]: you also need to remove it from https://github.com/rust-lang/rust/blob/56237d75b4271a8a2e0f47d86ea76ebf6d966152/src/librustc_mir/transform/check_consts/validation.rs#L337

Comment (oli-obk) [reply src/librustc_mir/transform/qualify_consts.rs]: and there's a test: https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-eval/assign-to-static-within-other-static.rs#L10

**Comment by Centril** [comment]
I'm not sure what discussions there have been here, but as with https://github.com/rust-lang/rust/pull/66302, I think it is important that this PR come with some descriptions, rationales, etc. Since the tests are passing, is this just a refactoring or does this change what code is allowed on stable?

File: src/librustc_mir/transform/qualify_consts.rs (hunk 1 above)
Comment (spastorino) [reply src/librustc_mir/transform/qualify_consts.rs]: Added the missing bits.

**Comment by oli-obk** [comment]
> Since the tests are passing, is this just a refactoring or does this change what code is allowed on stable?

This is just a refactoring. As the removed code itself said, it only a heuristic catching a few cases early instead of leaving it all to const eval. It's easy to work around the static check and then run into the miri-engine check.

**Pull request merged**